### DISCUSSION TITLE
[Functions] 502 SyntaxError: Cannot use import statement outside a module

### DISCUSSION THREAD

**Discussion by markusand** [vercel/vercel#5685] [Help]
I've build a simple project with functions. Everything works perfect on dev environment, but fails on production.
From logs the issue is when loading axios, but I don't understand the reason.

```
[GET] /api/events
01:39:41:06
2021-01-12T00:39:41.302Z	undefined	ERROR	/var/task/slack-bot/index.js:1
import axios from 'axios';
^^^^^^
SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:931:16)
    at Module._compile (internal/modules/cjs/loader.js:979:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1035:10)
    at Module.load (internal/modules/cjs/loader.js:879:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:903:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/var/task/api/events.js:1:1)
    at Module._compile (internal/modules/cjs/loader.js:1015:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1035:10)
RequestId: 14400678-8f2b-4ed2-8c30-411dc9f8eb02 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

**Comment by davidhartsough** [comment]
I wish I could suggest an answer but this problem just happened to me as well. I have had serverless functions working perfectly fine in production up until today when I did another deploy. Now I'm getting the same error as you. Mine also occurs on an axios import (although I use imports everywhere). 

```
[OPTIONS] /api/auth/refresh
17:05:18:39
2021-02-03T23:05:18.996Z	undefined	ERROR	/var/task/utils.js:1
import axios from "axios";
^^^^^^
SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/var/task/api/auth/refresh.js:3:1)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
RequestId: 504600a1-65fa-4d8c-8ce5-f1aa87efe84d Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

I tried adding this to my package.json: 
```
"type": "module",
```
But that lead to this error: 
```
[OPTIONS] /api/auth/refresh
16:31:12:71
2021-02-03T22:31:13.245Z	
undefined	
ERROR	
Uncaught Exception 	
{
   "errorType":"Error",
   "errorMessage":"Must use import to load ES Module: /var/task/___vc_launcher.js\nrequire() of ES modules is not supported.\nrequire() of /var/task/___vc_launcher.js from /var/runtime/UserFunction.js is an ES module file as it is a .js file whose nearest parent package.json contains \"type\": \"module\" which defines all .js files in that package scope as ES modules.\nInstead rename ___vc_launcher.js to end in .cjs, change the requiring code to use import(), or remove \"type\": \"module\" from /var/task/package.json.\n",
   "code":"ERR_REQUIRE_ESM",
   "stack":[
      "Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /var/task/___vc_launcher.js",
      "require() of ES modules is not supported.",
      "require() of /var/task/___vc_launcher.js from /var/runtime/UserFunction.js is an ES module file as it is a .js file whose nearest parent package.json contains \"type\": \"module\" which defines all .js files in that package scope as ES modules.",
      "Instead rename ___vc_launcher.js to end in .cjs, change the requiring code to use import(), or remove \"type\": \"module\" from /var/task/package.json.",
      "",
      "    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1015:13)",
      "    at Module.load (internal/modules/cjs/loader.js:863:32)",
      "    at Function.Module._load (internal/modules/cjs/loader.js:708:14)",
      "    at Module.require (internal/modules/cjs/loader.js:887:19)",
      "    at require (internal/modules/cjs/helpers.js:74:18)",
      "    at _tryRequire (/var/runtime/UserFunction.js:75:12)",
      "    at _loadUserApp (/var/runtime/UserFunction.js:95:12)",
      "    at Object.module.exports.load (/var/runtime/UserFunction.js:140:17)",
      "    at Object.<anonymous> (/var/runtime/index.js:43:30)",
      "    at Module._compile (internal/modules/cjs/loader.js:999:30)"
   ]
}
Unknown application error occurred
Error
```

So that's not the solution. 
I'd really love some help on this too. 
None of my serverless functions are working now. (My whole API and website are down due to this. Yikes!)

**Reply by walpolea** [reply]
Have the same issue, but not with axios, just a separate local js file. It was working great, then I deployed some changes and it's all broken as described above.

**Reply by davidhartsough** [reply]
I know this isn't a great solution for everyone, but I thought I should go back and tell people what worked for me. If you can do the same, maybe it will help you to: 

1. Switch to TypeScript.
2. Use Node version 14.x
3. If any import errors are still occurring, try to isolate the imported libraries by moving them to separate modules, testing out different arrangements of imports. For example, make a separate module like this: 

```typescript
import axios from "axios";

export async function get(path: string): Promise<any> {
  const { data } = await axios.get(path);
  return data;
}
```
(Or something similar that you can then import into a serverless function to see where exactly Vercel throws the error.)

I didn't receive any errors after switching to TypeScript and Node version 14.x. (But I should note here that back in January/February, choosing between Node versions [12 or 14] wasn't an option in Vercel.)

**Comment by tonprince** [comment]
I have the same issue, tried both module and commonjs in package.json without success. I always got the following error:

`2021-06-07T03:57:19.207Z	undefined	ERROR	Uncaught Exception 	{"errorType":"Error","errorMessage":"Must use import to load ES Module: /var/task/___vc_launcher.js\nrequire() of ES modules is not supported.\nrequire() of /var/task/___vc_launcher.js from /var/runtime/UserFunction.js is an ES module file as it is a .js file whose nearest parent package.json contains \"type\": \"module\" which defines all .js files in that package scope as ES modules.\nInstead rename ___vc_launcher.js to end in .cjs, change the requiring code to use import(), or remove \"type\": \"module\" from /var/task/package.json.\n","code":"ERR_REQUIRE_ESM","stack":["Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /var/task/___vc_launcher.js","require() of ES modules is not supported.","require() of /var/task/___vc_launcher.js from /var/runtime/UserFunction.js is an ES module file as it is a .js file whose nearest parent package.json contains \"type\": \"module\" which defines all .js files in that package scope as ES modules.","Instead rename ___vc_launcher.js to end in .cjs, change the requiring code to use import(), or remove \"type\": \"module\" from /var/task/package.json.","","    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1085:13)","    at Module.load (internal/modules/cjs/loader.js:933:32)","    at Function.Module._load (internal/modules/cjs/loader.js:774:14)","    at Module.require (internal/modules/cjs/loader.js:957:19)","    at require (internal/modules/cjs/helpers.js:88:18)","    at _tryRequire (/var/runtime/UserFunction.js:75:12)","    at _loadUserApp (/var/runtime/UserFunction.js:95:12)","    at Object.module.exports.load (/var/runtime/UserFunction.js:140:17)","    at Object.<anonymous> (/var/runtime/index.js:43:30)","    at Module._compile (internal/modules/cjs/loader.js:1068:30)"]}Unknown application error occurredError`

**Comment by Shubham-Kumar-2000** [comment]
Hi did anyone got a solution?

**Reply by Shubham-Kumar-2000** [reply]
For those who are still facing this error:

Adding `{ "src": "package.json", "use": "@vercel/node" }` to builds in now.json.
And replacing `@now/node` with `@vercel/node` solved it for me.

**Reply by leerob** [reply]
You shouldn't even need to use `@vercel/node` in you `vercel.json` - it will automatically configure that for you!

**Reply by Shubham-Kumar-2000** [reply]
Ohh that's great, I didn't knew that.

**Comment by leerob** [comment]
Most of these errors seem related to ES Modules. I would verify that the package you're importing is publishing an ES Modules version, such that it can be used with `import`. 

If you're using Next.js (and functions inside Next.js), you can also turn on the experimental ESM support: https://nextjs.org/11-1

Also: You don't need to use `axios` with Next.js, because Next.js polyfills node-fetch for you: https://nextjs.org/docs/basic-features/supported-browsers-features#server-side-polyfills

**Reply by Shubham-Kumar-2000** [reply]
Surprisingly that was not the issue.
As far as I can guess it was because I was using `@now/node` which was deprecated long ago.

**Reply by leerob** [reply]
https://github.com/vercel/vercel/discussions/5685#discussioncomment-1202231

**Comment by joa-queen** [comment]
It seems this is happening again. At least started to happening to me on new deploys (old ones works just fine).

Worth mentioning I didn't change any of the code regarding the functions, it fails by just creating a new deploy (with changes somewhere else other than the functions failing).

```
import axios from 'axios';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1031:15)
    at Module._compile (node:internal/modules/cjs/loader:1065:27)
    at Module.m._compile (/Users/joaquin/.nvm/versions/node/v16.13.1/lib/node_modules/vercel/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/joaquin/.nvm/versions/node/v16.13.1/lib/node_modules/vercel/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
Error: Function `api/thread.ts` failed with exit code 1
```

I'm not using Next.js. My `vercel.json` looks like this:
```
{
  "version": 2,
  "routes": [
    { "handle": "filesystem" },
    { "src": "/\\+(.*)/(.*)", "dest": "/api/thread.ts?communitySlug=$1&threadSlug=$2" },
    { "src": "/.*", "dest": "/index.html" }
  ]
}
```

Any advice on how to solve this? It's preventing me to deploy new versions of the site.

**Reply by Ehesp** [reply]
Also seeing this. Only have a small function and have converted it to pure JS for now... couldn't figure it out.

**Reply by joa-queen** [reply]
@leerob this looks like a regression?

**Reply by CODPHISH** [reply]
i have the same issue, downgrade vercel/cli to 28.10.2, everything works fine on dev, but if use higher version of vercel/cli or just deploy on cloud which will use vercel/cli latest version, and will definitly cause  "SyntaxError: Cannot use import statement outside a module"

**Reply by Andysimmonsisme** [reply]
@CODPHISH THANK YOU. I've been beating my head against a wall trying to figure out why this once worked and now it doesn't. Downgraded to 28.10.2 and it works just fine. This is super frustrating that this happens and you'd think there would be more descriptive errors for it!

**Reply by Ipicon** [reply]
after being stuck on this for hours and refusing to downgrade because it was not intuitive, downgraded to 28.10.12 and the problem doesn't occur anymore.

**Comment by peterhvw** [comment]
For my use case I found a possible solution. 
In my project I'm using the vercel.json to point to an api directory with some serverless function handlers.
```json
{
    "$schema": "https://openapi.vercel.sh/vercel.json",
    "framework": "nuxtjs",
    "functions": {
         "api/mail.ts": {
 	     "memory": 1024,
 	     "maxDuration": 10
 	}	
    }
}
```
I faced the same issue as described above. Somehow, a different project host on vercel didn't have this issues. 
The difference between the project was how the tsconfig.json was configured. 

To solve the issues i use the "excludeFiles" and "includeFiles" properties to point to a different tsconfig.json file. 
I placed the new tsconfig.json in the api directory. 
```json
{
    "$schema": "https://openapi.vercel.sh/vercel.json",
    "functions": {
        "api/mail.ts": {
            "memory": 1024,
            "maxDuration": 10,
            "excludeFiles": "./tsconfig.json",
            "includeFiles": "./api/tsconfig.json"
        }
    }
}
```
The tsconfig.json i used in the api folder: 
```json
{
    "compilerOptions": 
    {
        "target": "ES2020",
        "module": "commonjs",
        "lib": ["ES2020"],
        "strict": true,
        "noImplicitAny": true,
        "strictNullChecks": true,
        "strictFunctionTypes": true,
        "strictBindCallApply": true,
        "strictPropertyInitialization": true,
        "noImplicitThis": true,
        "alwaysStrict": true,
        "esModuleInterop": true
    }
 }
```

**Comment by nicolasconnault** [comment]
It's July 2023 and I just came across this issue. 
The simple solution for me was to replace api/serverless.ts with api/serverless.js:

```
const dotenv = require("dotenv");
const appModule = require("../src/app");

dotenv.config();

// Require the framework
const Fastify = require("fastify");

// Instantiate Fastify with some config
const appWrapper = Fastify({
  logger: false,
});

// Register your application as a normal plugin.
appWrapper.register(appModule, {
    prefix: '/'
});

module.exports = async function(req, res) {
  await appWrapper.ready();
  appWrapper.server.emit('request', req, res);
}
```

Don't forget to update your vercel.json file:
```
{
  "rewrites": [
      {
          "source": "/(.*)",
          "destination": "/api/serverless.js"
      }
  ]
}
```

This now works both locally with vercel dev and as a serverless endpoint

**Reply by ReddixT** [reply]
what if you dont want to use js instead of ts?

**Comment by Njitram2000** [comment]
For me the solution was very simple: I had an angular project whose tsconfig.json automatically had `"module": "es2022",` set. However, the vercel API requires this to be set to `"module": "commonjs"`

The Vercel API shares the TS settings with Angular which may be unfortunate. I haven't determined yet if this will cause issues for my actual angular app later down the line.

What I do find odd is that, from what I understand, commonJs uses `require` while es modules use `import`. So it seems like Vercel should want ES modules rather than commonJs...

**Update** after some help from ChatGPT (added to warn about the source 😉):
Angular runs in the browser so the module type is ES6 (native to browsers) while the serverless functions run in NodeJs so it needs commonJs (native to NodeJs).
The trick is to add a tsconfig.json to the api folder that contains the serverless functions .ts files. The tsconfig.json is then dedicated to the angular app while Vercel wil use the one in the api folder purely for the serverless functions.

**Comment by TechQuery** [comment]
Hey guys! Resolution is here: https://github.com/vercel/vercel/discussions/10317#discussioncomment-6787728
